src/libstd/sys/unix/fs.rs

```diff
@@ -511,15 +522,15 @@ pub fn lstat(p: &Path) -> io::Result<FileAttr> {
 
 pub fn canonicalize(p: &Path) -> io::Result<PathBuf> {
     let path = try!(CString::new(p.as_os_str().as_bytes()));
-    let mut buf = vec![0u8; 16 * 1024];
+    let buf;
     unsafe {
-        let r = c::realpath(path.as_ptr(), buf.as_mut_ptr() as *mut _);
+        let r = c::realpath(path.as_ptr(), ptr::null_mut());
         if r.is_null() {
             return Err(io::Error::last_os_error())
         }
+        buf = CStr::from_ptr(r).to_bytes().to_vec();
+        libc::free(r as *mut _);
```

Comment:
Do you think it might be a good idea to place an RAII guard on this? I believe `CStr::from_ptr(r).to_bytes().to_vec() will not panic currently, but if it changes the behavior in the future (e.g. `to_vec()` panics rather than aborting in case of OOM), the manually allocated memory will be leaked. Do we have to be concerned about this kind of situations? How much overhead an RAII guard incur?